HC-913 setup() runs twice (#61)
arduino/help-center-content: content/Software and Downloads/IDE Settings/If-setup-runs-twice.md

---

> title: "If setup() runs twice"

Reviewer: ```suggestion
title: "If your board runs the sketch twice"
```
Technically, we're not dealing with `setup()` specifically. The sketch will just run twice and can easily enter `loop()` the first time before resetting.

> title: "If setup() runs twice"

Reviewer: ```suggestion
title: "If your board runs the sketch twice"
```
Technically, we're not dealing with `setup()` specifically. The sketch will just run twice and can easily enter `loop()` the first time before resetting.

---

> The issue of the ```setup()``` function running twice when uploading a sketch has been reported in some boards.
> 
> Since a port can only be connected to one application at a time, the Arduino IDE has to close and open the port a number of times to upload a sketch while the Serial Monitor is open. This process resets the board and the program runs again.

Reviewer: ```suggestion
On some boards, uploading while the Serial Monitor is open causes the board to run once, before resetting and running the new sketch. This can cause `Serial.println()` output to display twice in the Serial Monitor.

This happens on boards without native USB capabilities (such as the UNO and Mega), which automatically reset when the port is opened. Since Arduino IDE needs to close the Serial Monitor for the upload tool to access the port, we get the following sequence of events:

1. Upload is started in Arduino IDE.
2. The Serial Monitor port is closed.
3. The port is opened by the upload tool. The board resets, and runs briefly runs any sketch currently on the board. 
4. After a short while (typically ~100 ms), upload finishes.
5. Serial Monitor reopens the port. The board resets a second time, and starts running the new sketch.
```

> The issue of the ```setup()``` function running twice when uploading a sketch has been reported in some boards.
> 
> Since a port can only be connected to one application at a time, the Arduino IDE has to close and open the port a number of times to upload a sketch while the Serial Monitor is open. This process resets the board and the program runs again.

Reviewer: ```suggestion
On some boards, uploading while the Serial Monitor is open causes the board to run once, before resetting and running the new sketch. This can cause `Serial.println()` output to display twice in the Serial Monitor.

This happens on boards without native USB capabilities (such as the UNO and Mega), which automatically reset when the port is opened. Since Arduino IDE needs to close the Serial Monitor for the upload tool to access the port, we get the following sequence of events:

1. Upload is started in Arduino IDE.
2. The Serial Monitor port is closed.
3. The port is opened by the upload tool. The board resets, and runs briefly runs any sketch currently on the board. 
4. After a short while (typically ~100 ms), upload finishes.
5. Serial Monitor reopens the port. The board resets a second time, and starts running the new sketch.
```

> The repeated output can be avoided in two ways

Reviewer: ```suggestion
The repeated output can be avoided in two ways:
```

> The repeated output can be avoided in two ways

Reviewer: ```suggestion
The repeated output can be avoided in two ways:
```

> 1. If one simply wants to repeatedly run the sketch to observe the output, there’s no need to upload the (same) sketch multiple times. Just press the reset button with the Serial Monitor open. This resets the board without closing the port, resulting in no repeated output.

Reviewer: ```suggestion
1. If one simply wants to repeatedly run the sketch to observe the output, there’s no need to upload the (same) sketch multiple times. Just press the reset button with the Serial Monitor open.
```
Actually the port will close and reopen, but only one time.

> 1. If one simply wants to repeatedly run the sketch to observe the output, there’s no need to upload the (same) sketch multiple times. Just press the reset button with the Serial Monitor open. This resets the board without closing the port, resulting in no repeated output.

Reviewer: ```suggestion
1. If one simply wants to repeatedly run the sketch to observe the output, there’s no need to upload the (same) sketch multiple times. Just press the reset button with the Serial Monitor open.
```
Actually the port will close and reopen, but only one time.

> 2. The board resets after  approximately 100 milliseconds. By adding ```delay(150)``` at the start of the ```setup()``` function, the board will be reset before any of the following code is executed, preventing the repeated output.

Reviewer: ```suggestion
2. The board resets after approximately 100 milliseconds. By adding `delay(150)` at the start of the `setup()` function, the board will be reset before any of the following code is executed. Increase the delay if needed.
```

> 2. The board resets after  approximately 100 milliseconds. By adding ```delay(150)``` at the start of the ```setup()``` function, the board will be reset before any of the following code is executed, preventing the repeated output.

Reviewer: ```suggestion
2. The board resets after approximately 100 milliseconds. By adding `delay(150)` at the start of the `setup()` function, the board will be reset before any of the following code is executed. Increase the delay if needed.
```
